metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/ParserRunner.java

```diff
@@ -0,0 +1,234 @@
+    }
+    initializeStellar(client, parserConfigSupplier);
+    initializeParsers(parserConfigSupplier);
+  }
+
+  protected void initializeStellar(CuratorFramework client, Supplier<ParserConfigurations> parserConfigSupplier) {
+    Map<String, Object> cacheConfig = new HashMap<>();
+    for (String sensorType: sensorTypes) {
+      SensorParserConfig config = parserConfigSupplier.get().getSensorParserConfig(sensorType);
+
+      if (config != null) {
+        cacheConfig.putAll(config.getCacheConfig());
+      }
+    }
+    Cache<CachingStellarProcessor.Key, Object> cache = CachingStellarProcessor.createCache(cacheConfig);
+
+    Context.Builder builder = new Context.Builder()
+            .with(Context.Capabilities.ZOOKEEPER_CLIENT, () -> client)
+            .with(Context.Capabilities.GLOBAL_CONFIG, () -> parserConfigSupplier.get().getGlobalConfig())
+            .with(Context.Capabilities.STELLAR_CONFIG, () -> parserConfigSupplier.get().getGlobalConfig())
+            ;
+    if(cache != null) {
+      builder = builder.with(Context.Capabilities.CACHE, () -> cache);
+    }
+    this.stellarContext = builder.build();
+    StellarFunctions.initialize(stellarContext);
+  }
+
+  private void initializeParsers(Supplier<ParserConfigurations> parserConfigSupplier) {
+    sensorToParserComponentMap = new HashMap<>();
+    for(String sensorType: sensorTypes) {
+      if (parserConfigSupplier.get().getSensorParserConfig(sensorType) == null) {
+        throw new IllegalStateException(String.format("Could not initialize parsers.  Cannot find config for sensor %s.",
+                sensorType));
+      }
+
+      SensorParserConfig parserConfig = parserConfigSupplier.get().getSensorParserConfig(sensorType);
+      // create message parser
+      MessageParser<JSONObject> parser = ReflectionUtils
+              .createInstance(parserConfig.getParserClassName());
+
+      // create message filter
+      MessageFilter<JSONObject> filter = null;
+      parserConfig.getParserConfig().putIfAbsent("stellarContext", stellarContext);
+      if (!StringUtils.isEmpty(parserConfig.getFilterClassName())) {
+        filter = Filters.get(
+                parserConfig.getFilterClassName(),
+                parserConfig.getParserConfig()
+        );
+      }
+      parser.configure(parserConfig.getParserConfig());
+      parser.init();
+      sensorToParserComponentMap.put(sensorType, new ParserComponent(parser, filter));
+    }
+  }
+
+  public void execute(String sensorType, RawMessage rawMessage, ParserConfigurations parserConfigurations) {
+    if (onSuccess == null) {
+      throw new IllegalStateException("An onSuccess function must be set before parsing a message.");
+    }
+    if (onError == null) {
+      throw new IllegalStateException("An onError function must be set before parsing a message.");
+    }
+    SensorParserConfig sensorParserConfig = parserConfigurations.getSensorParserConfig(sensorType);
+    if (sensorParserConfig != null) {
```

Comment:
Why have the api return the List AND do the callback.  Just have the callback.

Reply:
It's going to be one or the other.  Either it's synchronous and we return a list or it's asynchronous and we expose callbacks.  The interface would have to be different in each case.